File: Graph/Evaluate_Division.py
#Problem Link:- https://leetcode.com/problems/evaluate-division/

"""you are given an array of variable pairs equations and an array of real numbers values, 
where equations[i] = [Ai, Bi] and values[i] represent the equation Ai / Bi = values[i]. Each 
Ai or Bi is a string that represents a single variable.You are also given some queries, where 
queries[j] = [Cj, Dj] represents the jth query where you must find the answer for Cj / Dj = ?.
Return the answers to all queries. If a single answer cannot be determined, return -1.0."""

class Solution(object):
    def calcEquation(self, equations, values, queries):
        from collections import defaultdict
        graph = defaultdict(dict)
        for (i,j),val in zip(equations,values):
            graph[i][j] = val
            graph[j][i] = 1.0/val
        res = []
        
        def dfs(x,y,visited):
            if x not in graph or y not in graph:
                return -1.0
            if y in graph[x]:
                return graph[x][y]
            
            for i in graph[x]:
                if i not in visited:
                    visited.add(i)
                    temp = dfs(i, y, visited)
                    if temp == -1:
                        continue
                    else:
                        return graph[x][i] * temp
            return -1
        for query in queries:
            res.append(dfs(query[0], query[1], set()))
        return res

File: Graph/EvaluateDivision.cpp
//Problem Link:- https://leetcode.com/problems/evaluate-division/

#include<bits/stdc++.h>
using namespace std;

class Solution {
    private:
        unordered_map<string,unordered_map<string,double>> graph;
public:
    double dfs(string start, string dest,unordered_set<string> &vis){
        if (graph.find(start) == graph.end()){
            return -1;
        }
        if(graph[start].find(dest)!=graph[start].end())
            return graph[start][dest];
        vis.insert(start);
        for(auto i:graph[start])
        {
            //we dont want to revisit our previously visited strings
            if(!vis.count(i.first))
            {
                double ans=dfs(i.first,dest,vis);
                if(ans!=-1.0)
                    return (double)ans * (i.second);
            }       
        }
        return -1.0;
    }
    
    vector<double> calcEquation(vector<vector<string>>& eq, vector<double>& values, vector<vector<string>>& queries) {
        for(int i=0;i<eq.size();i++){
            string a = eq[i][0];
            string b = eq[i][1];
            double val = values[i];
            graph[a].insert({b,val});
            graph[b].insert({a,(double)1/val});
        }
        vector<double> res;
        for(int i=0;i<queries.size();i++){
            unordered_set<string> vis;
            double ans = dfs(queries[i][0],queries[i][1],vis);
            res.push_back(ans);
        }
        return res;
    }
};